authorization/src/main/kotlin/org/modelix/authorization/AuthorizationPlugin.kt

```diff
@@ -0,0 +1,242 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.authorization
+
+import com.auth0.jwt.JWT
+import com.auth0.jwt.algorithms.Algorithm
+import com.google.common.cache.CacheBuilder
+import io.ktor.http.HttpStatusCode
+import io.ktor.server.application.Application
+import io.ktor.server.application.ApplicationCall
+import io.ktor.server.application.ApplicationCallPipeline
+import io.ktor.server.application.BaseRouteScopedPlugin
+import io.ktor.server.application.call
+import io.ktor.server.application.install
+import io.ktor.server.auth.Authentication
+import io.ktor.server.auth.AuthenticationContext
+import io.ktor.server.auth.AuthenticationProvider
+import io.ktor.server.auth.authenticate
+import io.ktor.server.auth.jwt.jwt
+import io.ktor.server.auth.principal
+import io.ktor.server.html.respondHtml
+import io.ktor.server.plugins.forwardedheaders.XForwardedHeaders
+import io.ktor.server.response.respond
+import io.ktor.server.response.respondText
+import io.ktor.server.routing.Route
+import io.ktor.server.routing.application
+import io.ktor.server.routing.get
+import io.ktor.server.routing.routing
+import io.ktor.util.AttributeKey
+import org.modelix.authorization.permissions.AccessControlData
+import org.modelix.authorization.permissions.PermissionEvaluator
+import org.modelix.authorization.permissions.PermissionId
+import org.modelix.authorization.permissions.PermissionParts
+import org.modelix.authorization.permissions.SchemaInstance
+import org.modelix.authorization.permissions.UserId
+import java.nio.charset.StandardCharsets
+import java.util.Base64
+import java.util.Collections
+import java.util.concurrent.TimeUnit
+
+/**
+ * JWT based authorization plugin.
+ */
+object ModelixAuthorization : BaseRouteScopedPlugin<IModelixAuthorizationConfig, ModelixAuthorizationPluginInstance> {
+    override fun install(
+        pipeline: ApplicationCallPipeline,
+        configure: IModelixAuthorizationConfig.() -> Unit,
+    ): ModelixAuthorizationPluginInstance {
+        val config = ModelixAuthorizationConfig().apply(configure)
+        val application = when (pipeline) {
+            is Route -> pipeline.application
+            is Application -> pipeline
+            else -> error("Unsupported pipeline type: ${pipeline::class}")
+        }
+        val installedIntoRoute = pipeline as? Route
+
+        application.install(XForwardedHeaders)
+        application.install(Authentication) {
+            if (config.shouldGenerateFakeTokens()) {
```

Comment:
Has the whole fake token thing a use besides unit tests? If not, can't we find a way that doesn't expose this (potentially dangerous -- might accidentally open up routes) option to the user?

Reply:
The generated token doesn't grant any permissions. It just provides a user name.

Reply:
But it's only used for unit testing purposes?

Reply:
It's used whenever the server runs without configured authorization. That's the previous behavior. I didn't want to change it.